### PR TITLE
CubeFS Project Security Self-Assessment - Security Pals

### PR DESCRIPTION
**Pull request by avinashnarasimhan18**
Created and added the first draft of the CubeFS Project Security Self-Assessment. Please feel free to share your feedback on the security self-assessment.

### CODE REVIEW

File: assessments/projects/cubefs/self-assessment.md [hunk 1]
```diff
@@ -0,0 +1,300 @@
...
+
+## Metadata
+
+|   |  |
+| -- | -- |
+| Software | https://github.com/cubefs/cubefs  |
+| Website | https://cubefs.io/ |
+| Security Provider | No  |
+| Languages | Golang |
+| SBOM | [CubeFS SBOM File](https://github.com/avinashnarasimhan18/tag-security/blob/4e6b73248f7807cef59a456ab87d8f357094f380/assessments/projects/cubefs/cubefs_sbom.json)|
```
Comment (ragashreeshekar) [inline assessments/projects/cubefs/self-assessment.md:40]: This link seems to be from your personal fork of the repository.
Please update it with the link from the project.

**Comment by avinashnarasimhan18** [comment]
@eddie-knight Thank you for starting the review on the self-assessment and giving us some initial comments.

CubeFS has not made the SBOM publicly available, so our team generated one using the FOSSA CLI. Would you prefer we keep the SBOM or remove it and mention that it is not publicly available? Please let us know.

**Comment by avinashnarasimhan18** [comment]
@ragashreeshekar We have started working on your review comments. Please let us know if there are any other changes to be made. We will change accordingly and push all of them in one commit.

File: assessments/projects/cubefs/self-assessment.md [hunk 2]
```diff
@@ -0,0 +1,300 @@
...
+* **User Authentication and Authorization** - 
+  * Actors: Authentication Service, CubeFS Components (e.g., Master, Metadata, DataNodes)
+  * Actor Interaction: Authentication Service validates user credentials via service requests and grants access permissions to CubeFS components based on defined policies
+
+* **Volume Access Control** -
+  * Actors: Volume Management System, Container/Client Applications
+  * Actor Interaction: Volume Management System manages access permissions and interacts with container/client applications to grant or restrict access to volumes. Regular audits to ensure access permissions align with security policies. 
+
+
+### Goals
```
Comment (ragashreeshekar) [inline assessments/projects/cubefs/self-assessment.md:106]: Goals section usually talks about what the project aims to accomplish (could incl. security goals but that isn't the main focus). A good example is https://github.com/cncf/tag-security/blob/main/assessments/projects/kyverno/self-assessment.md#goal

File: assessments/projects/cubefs/self-assessment.md [hunk 3]
```diff
@@ -0,0 +1,300 @@
...
+
+  * **Goal**: Ensure security measures do not compromise system performance.
+  * **Security Guarantee**: Continuous refinement of security mechanisms (e.g., optimizing writeback cache, addressing performance limitations in encryption) without compromising overall system performance.
+
+* **Key Management and Rotation**:
+
+  * **Goal**: Safeguard shared keys and support regular key rotation for enhanced security.
+  * **Security Guarantee**: Future enhancements to support key rotation, reducing the risk of key compromise and strengthening encryption mechanisms.
+
+### Non-goals
```
Comment (ragashreeshekar) [inline assessments/projects/cubefs/self-assessment.md:143]: Non-goals section usually talks about what actions are out of scope for the same. A good example is https://github.com/cncf/tag-security/blob/main/assessments/projects/kyverno/self-assessment.md#non-goals

File: assessments/projects/cubefs/self-assessment.md [hunk 4]
```diff
@@ -0,0 +1,300 @@
...
+  * Actor Interaction: The Metadata Subsystem will update and manage the metadata shards while coordinating with the Master Node for consistency and replication
+
+* **User Authentication and Authorization** - 
+  * Actors: Authentication Service, CubeFS Components (e.g., Master, Metadata, DataNodes)
+  * Actor Interaction: Authentication Service validates user credentials via service requests and grants access permissions to CubeFS components based on defined policies
+
+* **Volume Access Control** -
+  * Actors: Volume Management System, Container/Client Applications
+  * Actor Interaction: Volume Management System manages access permissions and interacts with container/client applications to grant or restrict access to volumes. Regular audits to ensure access permissions align with security policies. 
+
```
Comment (ragashreeshekar) [inline assessments/projects/cubefs/self-assessment.md:104]: What are the components involved and workflow in the Data read operations?

File: assessments/projects/cubefs/self-assessment.md [hunk 5]
```diff
@@ -0,0 +1,300 @@
...
+* **Elimination of All Operational Errors** - CubeFS does not aim to eradicate all potential errors or misconfigurations during operational usage.
+
+* **Providing Protection Beyond Defined Use Cases** - CubeFS does not extend security protections to scenarios or use cases not explicitly defined or supported. CubeFS focuses on securing the defined use cases and may not cover security aspects outside these boundaries.
+
+## Self-assessment use
+
+This self-assessment is created to perform an internal analysis of the project's security.  It is not intended to provide a security audit of CubeFS, or function as an independent assessment or attestation of CubeFS's security health. This document serves to provide CubeFS users with an initial understanding of CubeFS's security, where to find existing security documentation, their plans for security, and a general overview of its security practices, both for the development as well as security of CubeFS. This document provides the CNCF TAG-Security with an initial understanding of the project and its security posture.
+
+## Security functions and features
```
Comment (ragashreeshekar) [inline assessments/projects/cubefs/self-assessment.md:163]: Considering the project is in cloud storage domain, pointers on data backup, recovery could be useful to include here.

File: assessments/projects/cubefs/self-assessment.md [hunk 6]
```diff
@@ -0,0 +1,300 @@
...
+
+* **Development Pipeline** - 
+
+  * **Version control** - The CubeFS repository uses Git as its primary version control tool.
+
+  * **Code contribution** - CubeFS has defined a document for users and other community members who want to contribute to the repository. The document contains development style guidelines and commit message guidelines. 
+
+  * **Commit Signing** - All commits have to be signed following certain guidelines and go through a DCO (Developer Certificate of Origin) check. 
+
+  * **CI/CD pipeline** - CubeFS uses automated CI/CD pipelines to build and test code and uses the Travis CI service to run all checks including unit and integration tests.
```
Comment (ragashreeshekar) [inline assessments/projects/cubefs/self-assessment.md:204]: Are there any SAST tools for container image vulnerabilities, static code analysis or source composition analysis included in their CI pipeline? Please include here.

**Comment by ragashreeshekar** [comment]
> @ragashreeshekar We have started working on your review comments. Please let us know if there are any other changes to be made. We will change accordingly and push all of them in one commit.

Thanks for the PR @avinashnarasimhan18 and team, appreciate the efforts. 
I have completed first pass of review. Please feel free to reach out here or on slack for any questions and clarifications. 
Along with addressing the comments, kindly update the PR branch with the latest content in the repo as this branch is out-of-date with the base branch.

**Comment by avinashnarasimhan18** [comment]
> > @ragashreeshekar We have started working on your review comments. Please let us know if there are any other changes to be made. We will change accordingly and push all of them in one commit.
> 
> Thanks for the PR @avinashnarasimhan18 and team, appreciate the efforts. I have completed first pass of review. Please feel free to reach out here or on slack for any questions and clarifications. Along with addressing the comments, kindly update the PR branch with the latest content in the repo as this branch is out-of-date with the base branch.

Thank you @ragashreeshekar. We will start working on your comments right away and update once it's done. We will ensure the branch is updated as well.

**Comment by avinashnarasimhan18** [comment]
@ragashreeshekar , me and my team have addressed your and Eddie's review comments. Please take a look and let us know if there are any other changes you would like us to make.

File: assessments/projects/cubefs/self-assessment.md [hunk 7]
```diff
@@ -0,0 +1,257 @@
...
+
+## Metadata
+
+|   |  |
+| -- | -- |
+| Software | https://github.com/cubefs/cubefs  |
+| Website | https://cubefs.io/ |
+| Security Provider | No  |
+| Languages | Golang |
+| SBOM | Currently, no SBOM is publicly available for CubeFS|
```
Comment (eddie-knight) [inline assessments/projects/cubefs/self-assessment.md:31]: Recommended minor change: highlight that target state is not yet achieved, while removing red-herring questions about a potential privately available SBOM.
 
```suggestion
| SBOM | CubeFS does not currently generate an SBOM on release |
```

Comment (avinashnarasimhan18) [reply assessments/projects/cubefs/self-assessment.md]: @eddie-knight, we have added the change you suggested. Please let us know if it's okay.

File: assessments/projects/cubefs/self-assessment.md [hunk 8]
```diff
@@ -0,0 +1,257 @@
...
+    * **Suggested Resolution** - To meet the highest standards of the OpenSSF Scorecard, and ensure maximum project integrity, a SLSA provenance file should be included in the assets for each release, which in turn will increase the Signed-Releases score. The details on this can be found on the SLSA [official site](https://slsa.dev/).
+ 
+* **Use of Static Application Security Testing (SAST) tool** -
+    * **Existing Issue** - CubeFS’s current commit procedure does not include the use of a Static Application Security Testing (SAST) tool. Though some commits use SAST tools, not all of them do. This could result in unsafe code, and potential security threats being merged with the main branch.
+    * **Suggested Resolution** - By standardizing the use of SAST tools before every commit within the pipeline, regardless of the commit’s purpose, CubeFS can ensure that the integrity of their code is consistently up to date. Details of SAST can be found [here](https://github.com/ossf/scorecard/blob/4edb07802fdad892fa8d10f8fd47666b6ccc27c9/docs/checks.md#sast).
```
Comment (eddie-knight) [inline assessments/projects/cubefs/self-assessment.md:255]:
```suggestion
```

Following discussions from the review team, we are asking that Security Pals please remove any recommendations from the scope of this document. You may optionally communicate recommendations for projects in a more accessible location, such as in a project community channel or GitHub Issue.

Comment (avinashnarasimhan18) [reply assessments/projects/cubefs/self-assessment.md]: @eddie-knight, per your suggestion, we have removed the action items section from this document. We will figure out how to communicate our suggestions to the CubeFS team.

File: assessments/projects/cubefs/self-assessment.md (hunk 1 above)
Comment (avinashnarasimhan18) [reply assessments/projects/cubefs/self-assessment.md]: The SBOM Section has been updated.

File: assessments/projects/cubefs/self-assessment.md [hunk 9]
```diff
@@ -0,0 +1,300 @@
...
+security of the project, determining gaps in their security, and preparing any security
+documentation for their users. This document is ideal for projects currently in the
+CNCF **sandbox** as well as projects that are looking to receive a joint assessment and
+currently in CNCF **incubation**.
+
+For a detailed guide with step-by-step discussion and examples, check out the free 
+Express Learning course provided by Linux Foundation Training & Certification: 
+[Security Assessments for Open Source Projects](https://training.linuxfoundation.org/express-learning/security-self-assessments-for-open-source-projects-lfel1005/).
+
+# Self-assessment outline
```
[the remark this answers is not in the document]
Comment (avinashnarasimhan18) [reply assessments/projects/cubefs/self-assessment.md]: Template text has been removed and additional heading has been removed.

File: assessments/projects/cubefs/self-assessment.md (hunk 2 above)
Comment (avinashnarasimhan18) [reply assessments/projects/cubefs/self-assessment.md:96]: The goals section has been updated as per the review comment.

File: assessments/projects/cubefs/self-assessment.md (hunk 3 above)
Comment (avinashnarasimhan18) [reply assessments/projects/cubefs/self-assessment.md:104]: The non-goals section has been updated as per the review comment.

File: assessments/projects/cubefs/self-assessment.md (hunk 4 above)
Comment (avinashnarasimhan18) [reply assessments/projects/cubefs/self-assessment.md:94]: The data read operation has been expounded upon.

File: assessments/projects/cubefs/self-assessment.md (hunk 6 above)
Comment (avinashnarasimhan18) [reply assessments/projects/cubefs/self-assessment.md]: Lack of SAST tools in CubeFS CI/CD pipeline has been addressed.

**Comment by ragashreeshekar** [comment]
Thank you for the updates. Looks good to me. 
Tagging @eddie-knight for your review, and @JustinCappos both for your opinion on including maintainers for their ack and review of this work, followed by chair review (as applicable) and merging them to the repo.

File: assessments/projects/cubefs/self-assessment.md [hunk 10]
```diff
@@ -0,0 +1,235 @@
+# CubeFS Self-assessment
+This Self-assessment document has been created by students at NYU Tandon to help provide information regarding CubeFS, determining gaps in their security, and preparing any security
```
Comment (JustinCappos) [inline assessments/projects/cubefs/self-assessment.md:2]: Do we want to change this lead in?  I would think that it should mention the Security Pals effort.

Comment (avinashnarasimhan18) [reply assessments/projects/cubefs/self-assessment.md:2]: @JustinCappos do you have any suggestions for this?

Comment (JustinCappos) [reply assessments/projects/cubefs/self-assessment.md:2]: Please use: 

This assessment was created by community members as part of the [Security Pals](https://github.com/cncf/tag-security/issues/1102) process, and is currently pending changes from the maintainer team.

Comment (avinashnarasimhan18) [reply assessments/projects/cubefs/self-assessment.md]: Thank you @JustinCappos! The suggested change has been implemented. Please review.

File: assessments/projects/cubefs/self-assessment.md [hunk 11]
```diff
@@ -0,0 +1,234 @@
+# CubeFS Self-assessment
+This Self-assessment document was created by community members as part of the [Security Pals](https://github.com/cncf/tag-security/issues/1102) process, and is currently pending changes from the maintainer team. It aims to help provide information regarding CubeFS, determining gaps in their security, and preparing any security documentation for their users.
```
Comment (ragashreeshekar) [inline assessments/projects/cubefs/self-assessment.md:2]:
```suggestion
This Self-assessment document was created by community members as part of the [Security Pals](https://github.com/cncf/tag-security/issues/1102) process, and is currently pending changes from the maintainer team. 

It aims to help provide information regarding CubeFS, determining gaps in their security, and preparing any security documentation for their users. 

```

File: assessments/projects/cubefs/self-assessment.md [hunk 12]
```diff
@@ -0,0 +1,237 @@
...
+* [Security functions and features](#security-functions-and-features)
+* [Project compliance](#project-compliance)
+* [Secure development practices](#secure-development-practices)
+* [Security issue resolution](#security-issue-resolution)
+* [Appendix](#appendix)
+
+## Metadata
+
+|   |  |
+| -- | -- |
```
Comment (ragashreeshekar) [inline assessments/projects/cubefs/self-assessment.md:27]:
```suggestion
| -- | -- |
| Assessment Stage | Incomplete | 
```

**Review by ragashreeshekar — APPROVED** [review]

File: assessments/projects/cubefs/self-assessment.md [hunk 13]
```diff
@@ -0,0 +1,238 @@
+# CubeFS Self-assessment
+This Self-assessment document was created by community members as part of the [Security Pals](https://github.com/cncf/tag-security/issues/1102) process, and is currently pending changes from the maintainer team. 
+
+It aims to help provide information regarding CubeFS, determining gaps in their security, and preparing any security documentation for their users.
```
Comment (torinvdb) [inline assessments/projects/cubefs/self-assessment.md:4]:
```suggestion
This Self-assessment document was created by community members as part of the [Security Pals](https://github.com/cncf/tag-security/issues/1102) process and is currently pending changes from the maintainer team. 

The intention of this assessment is to help provide security-related information, such as gap identification and user security considerations, that pertain to the CubeFS project.
```
@JustinCappos

**Pull request merged**